### PR TITLE
clone and use tree hash to get code instead of `Pkg.add`

### LINKED ISSUE #349: Automerge doesn't catch incorrect `subdir` parameters
hence https://github.com/JuliaRegistries/General/pull/30577. We could check this by `dev`ing the package (instead or in addition to `add`ing it):
```julia
(jl_4zbq4I) pkg> dev YisyAIFramework
     Cloning git-repo `https://github.com/SkyWorld117/YisyAIFramework.jl.git`
   Resolving package versions...
ERROR: expected package `YisyAIFramework [1833fadb]` to exist at path `/Users/eph/.julia/dev/YisyAIFramework/https:/github.com/SkyWorld117/YisyAIFramework.jl`
```
If `dev` doesn't throw, then at least it is a valid directory, though we probably should go one step further and check that the tree hash of the contents at that path matches the one in the registry. If anyone has any other ideas I'd be happy to hear it.

### PR DESCRIPTION
**Pull request by ericphanson**
* adds a non-optional `guideline_code_can_be_downloaded` check that cross-checks the tree hash from the PR's Versions.toml against the commit hash from the PR message body, and downloads the code to a `pkg_code_path` in the `GitHubAutoMergeData` object (replacing the `depot_path` that used to be there)
* The package can be `Pkg.add`'d and `import`ed steps now use a tempdir instead of the `depot_path` since we don't need to reuse the code anymore
* The license check now looks in the `pkg_code_path` to get the code
* breaking version number bump due to the requirement that PR bodies include the commit hash

closes #349

### CODE REVIEW

File: src/AutoMerge/util.jl [hunk 1]
```diff
@@ -26,6 +22,48 @@ function _clone_repo_into_dir(url::AbstractString, repo_dir)
...
+function load_files_from_url_and_tree_hash(destination::String, url::String, tree_hash::String)
+    pkg_clone_dir = mktempdir()
+    clone_success = try
+        _clone_repo_into_dir(url, pkg_clone_dir)
+        true
+    catch e
+        @error "Cloning $url failed"
+        false
+    end
+    Tar.extract(Cmd(`git archive $tree_hash`; dir=pkg_clone_dir), destination)
```
Comment (ericphanson) [inline src/AutoMerge/util.jl:42]: The fact that `Tar.extract` takes a `Cmd` as input is really nice!

Comment (DilumAluthge) [reply src/AutoMerge/util.jl:42]: @StefanKarpinski (for both Tar.jl and `Cmd` 😂)

**Comment by GunnarFarneback** [comment]
Looks good so far.

**Comment by ericphanson** [comment]
Added the hash code! Let's see what bors thinks

@bors try

**Comment by DilumAluthge** [comment]
You don't need the `@` before `bors`.

**Comment by ericphanson** [comment]
bors try

**Comment by ericphanson** [comment]
Oops, guess `Base.run` instead of `run` is needed afterall 😅. That was a confusing error message!

bors try

**Comment by ericphanson** [comment]
bors try

**Comment by ericphanson** [comment]
It looks like the tests might be failing here because we don't populate the PR bodies for the integration tests, e.g. https://github.com/bcbi-test/automerge-integration-test-repo/pull/7550

**Comment by ericphanson** [comment]
I think we could do something like make the download code check enable-able by an argument, then disable it for most of the integration tests (since many correspond to made-up versions anyway), but enable it for one that corresponds to actual commit, and pass in the correct commit in the PR body for that one. I'll try that...

**Comment by DilumAluthge** [comment]
> It looks like the tests might be failing here because we don't populate the PR bodies for the integration tests, e.g. [bcbi-test/automerge-integration-test-repo#7550](https://github.com/bcbi-test/automerge-integration-test-repo/pull/7550)

Let's add PR bodies for all the integration tests. The point of the integration tests is to replicate the real-world conditions as closely as possible.

**Comment by DilumAluthge** [comment]
> I think we could do something like make the download code check enable-able by an argument, then disable it for most of the integration tests (since many correspond to made-up versions anyway), but enable it for one that corresponds to actual commit, and pass in the correct commit in the PR body for that one. I'll try that...

Personally I'd rather not special case anything. Can we just do the "download code check" for all of the integration tests? IIRC, the Git tree SHAs in the integration tests all correspond to real commit in those repos.

**Comment by ericphanson** [comment]
> Let's add PR bodies for all the integration tests. The point of the integration tests is to replicate the real-world conditions as closely as possible.

I'm not sure it would add that much beyond what the tests I just added do, but it's a lot of configuration to add. With those we should already cover both cases (right commit and wrong commit).

bors try

**Comment by ericphanson** [comment]
> Personally I'd rather not special case anything. Can we just do the "download code check" for all of the integration tests? IIRC, the Git tree SHAs in the integration tests all correspond to real commit in those repos.

Sorry, I didn't see this before my last reply. I see, just the names and versions are made up then. That should be possible then, but I'm not sure it tests that much more than doing it once? It seems like we'd just be re-checking that the same commit + subdir pair corresponds to the same git-tree-sha many times, since the git-tree-sha's are reused throughout the tests.

bors try

**Comment by DilumAluthge** [comment]
My main concern was that it sounds like the integration tests don't reflect the way things are in production. So I figure if we just add PR bodies, then the integration tests run exactly the same way things do in production.

I don't mind running the same check over and over. What worries me is if the integration tests don't line up with production.

**Comment by ericphanson** [comment]
Ok, so I added the hashes so we test unconditionally. I haven't removed the `check_code_downloadable` option (though it defaults to true), though I think maybe I should? Wasn't sure if there could be some weird configuration problems on private registries and it would be nice to be able to disable it in that case, though in that case we should probably want to fix whatever the issue is anyway.

bors try

**Comment by DilumAluthge** [comment]
> Ok, so I added the hashes so we test unconditionally. I haven't removed the `check_code_downloadable` option (though it defaults to true), though I think maybe I should? Wasn't sure if there could be some weird configuration problems on private registries and it would be nice to be able to disable it in that case, though in that case we should probably want to fix whatever the issue is anyway.
> 
> bors try

Good question. It occurs to me that maybe some private registry setups might have firewalls that block Git or something?

So, how does this sound: leave the kwarg, but don't document, so it's private. Then we can remove it later if we want, without needing a breaking release.

**Comment by ericphanson** [comment]
Sounds good! This will be my last attempt tonight but I'll try to get back at it tomorrow.

bors try

**Comment by DilumAluthge** [comment]
Actually, I think we should do the following:
1. Remove the `check_code_downloadable` keyword argument.
2. Bump the major version number

My justification for the latter: right now, we only require that PRs have the package name and the version number. So people in private registries may be using AutoMerge with pull requests that are not opened by Registrator and thus do not have the commit in the PR body. This might make sense in a private registry setup, since you can give users direct merge permissions on the registry.

So then this is technically a breaking change since we now impose an additional constraint on the formatting of PRs.

**Comment by DilumAluthge** [comment]
You can bump the major version number and put `-DEV` at the end. That way, if we have more breaking changes to make in the short term, we can make them all in a single breaking release.

**Comment by GunnarFarneback** [comment]
> Good question. It occurs to me that maybe some private registry setups might have firewalls that block Git or something?

Trying to run automerge with a setup that blocks `git` would be a bit weird; the registry is typically stored in a `git` repository after all. A more likely complication is that a CI job running automerge might not be configured with sufficient keys or other credentials to access all package repos.

**Comment by ericphanson** [comment]
bors try

**Comment by ericphanson** [comment]
I don't know if it would be better or worse or no different for auth, but one option is we could switch the commandline invocations of `git` to https://github.com/JuliaVersionControl/Git.jl if we were willing to require Julia 1.6+.

**Comment by ericphanson** [comment]
bors try

**Comment by ericphanson** [comment]
Ok, should be good to do! I squashed the commits here.

bors try

**Review by DilumAluthge — APPROVED** [review]
Good to merge, I think.

**Review by GunnarFarneback — APPROVED** [review]

**Comment by ericphanson** [comment]
bors merge

**Pull request merged**